contracts/Permissions.sol

```diff
@@ -258,51 +234,59 @@ library Permissions {
         Enum.Operation operation,
         uint16 role
     ) public view {
-        bytes4 functionSig = bytes4(data);
-        // bools[0] = targetAllowed
-        // bools[1] = targetScoped
-        // bools[2] = sendAllowed
-        // bools[3] = functionAllowed
-        // bools[4] = functionscoped
-        // bools[5] = delegateAllowed
-        bool[6] memory bools;
-        bools[0] = self.roles[role].targetAddresses[targetAddress].allowed;
-        bools[1] = self.roles[role].targetAddresses[targetAddress].scoped;
-        bools[2] = self.roles[role].targetAddresses[targetAddress].sendAllowed;
-        bools[3] = self
-            .roles[role]
-            .targetAddresses[targetAddress]
-            .functions[functionSig]
-            .allowed;
-        bools[4] = self
-            .roles[role]
-            .targetAddresses[targetAddress]
-            .functions[functionSig]
-            .scoped;
-        bools[5] = self
-            .roles[role]
-            .targetAddresses[targetAddress]
-            .delegateCallAllowed;
-        if (data.length != 0 && data.length < 4) {
-            revert FunctionSignatureTooShort();
+        TargetAddress memory target = self.roles[role].targets[targetAddress];
+
+        // CLEARANCE: transversal - checks
+        if (value > 0 && !target.canSend) {
+            revert SendNotAllowed();
         }
-        if (operation == Enum.Operation.DelegateCall && !bools[5]) {
+
+        if (operation == Enum.Operation.DelegateCall && !target.canDelegate) {
             revert DelegateCallNotAllowed();
         }
-        if (!bools[0]) {
+
+        if (data.length != 0 && data.length < 4) {
+            revert FunctionSignatureTooShort();
+        }
+
+        // CLEARANCE: levels 1 2 and 3 - checks
+
+        /*
+         * For each address we have three clearance checks:
+         * Forbidden     - nothing was setup
+         * AddressPass   - all calls to this address are go, nothing more to check
+         * FunctionCheck - some functions on this address are allowed
+         */
+        bool isForbidden = target.clearance == AccessGranularity.NONE;
```

Comment:
Given reach of these variables is only used once, it didn't really make sense to store them. The if statements should just be a comparison against `target.clearance`. Although, this is very easy to read.

Reply:
Right

Reply:
Pushed